src/algebra/order_operation_defs.lean

```diff
@@ -0,0 +1,267 @@
+
+Since we want to have the freedom to apply them indepdently to
+
+* each relation `(≤)` and `(<)`,
+* each binary operation `(+)` and `(*)`,
+* left or right actions,
+
+we generalize the setup as follows.
+
+Start with two Types `M` and `N`, an "action" `μ : M → N → N` and a relation `r : N → N → Prop`.
+We introduce the two separate definitions
+
+* covariant:     `∀ (m : M) {n₁ n₂ : N}, r n₁ n₂ → r (μ m n₁) (μ m n₂)`,
+* contravariant: `∀ (m : M) {n₁ n₂ : N}, r (μ m n₁) (μ m n₂) → r n₁ n₂`,
+
+modeled on the monotonicity conditions above.
+
+We then introduce the 16 typeclasses obtained by choosing `M = N` and assuming
+
+* one of `covariant` and `contravariant`, with
+* the relation `r` being one of `(≤)` or `(<)`,
+* the operation `μ : M → M → M` being one of `(+)` or `(*)`, for left monotonicity,
+* the operation `μ : M → M → M` being one of `(function.swap (+))` or `(function.swap(*))`,
+  for right monotonicity,
+
+There are two main reasons for introducing several typeclasses.  The first, is that it is much more
+human-friendly to have a class called `has_add_le_add_right M` than one called
+`covariant M M (function.swap (+)) (≤)`.  The second, is that this way, we bound better the space
+for typeclass inference: e.g. we ask the typeclass search to look for `has_add` and `has_le`
+instances, instead of extending the search to all possible functions among different Types.
+
+The general approach is to formulate the lemma that you are interested and prove it, with the
+`ordered_[...]` typeclass of your liking.  After that, you convert the single typeclass,
+say `[ordered_cancel_monoid M]`, with three typeclasses, e.g.
+`[partial_order M] [left_cancel_semigroup M] [has_mul_le_mul_right M]`
+and have a go at seeing if the proof still works!
+
+Note that it is possible to combine several `co(ntra)variant` assumptions together.
+Indeed, the usual ordered typeclasses arise from assuming the pair
+`[has_mul_le_mul_left M] [has_lt_of_mul_le_mul_left M]` on top of order/algebraic assumptions.
+
+A formal remark is that normally the direct implication of `covariant` uses the `(≤)`-relation,
+while inverse implication of `contravariant` uses the `(<)`-relation. This need not be the case in
+general, but seems to be the most common usage. In the opposite direction, the implication
+
+```lean
+[semigroup M] [partial_order M] [has_le_of_mul_le_mul M] => left_cancel_semigroup α
+```
+holds (note the `has_le_[...]` as opposed to the more common idioms `has_lt_[...]` and
+`has_mul_le_[...]`.
+-/
+-- TODO: convert `has_exists_mul_of_le`, `has_exists_add_of_le`?
+-- TODO: relationship with add_con
+-- include equivalence of `left_cancel_semigroup` with
+-- `semigroup partial_order contravariant_class α α (*) (≤)`?
+-- use ⇒, as per Eric's suggestion?
+section covariants_and_contravariants
+
+variables {M N : Type*} (μ : M → N → N) (r : N → N → Prop) (m : M) {a b c : N}
+
+
+variables (M N)
```

Comment:
There's no point making these explicit, as they can be inferred from `μ` and `r`.

Reply:
Eric, this looks alright, until I feed into them `(*)`: is there a way to get Lean to understand the the `(*)` that I am giving it is the one on `N`, rather than some random other `*`?  See, for instance, line 122.  I could start annotating this as

```
@[to_additive]
lemma covariant_iff_covariant_mul [comm_semigroup N] :
  covariant ((*) : N → N → N) (r) ↔ covariant (function.swap (*) : N → N → N) (r) :=
```

but maybe the explicit variables are better?